### PR TITLE
implement data apis. add schema cache

### PR DESCRIPTION
**Pull request by iafuture**
This change implements the read/write paths in the python tester to make sure the HTTP proxy APIs are tested

### CODE REVIEW

File: src/k2/httpproxy/HTTPProxy.cpp [hunk 1]
```diff
@@ -476,92 +245,215 @@ HTTPProxy::_handleCreateSchema(sh::dto::CreateSchemaRequest&& request) {
...
+            }
+            return seastar::make_ready_future<std::tuple<Status, std::shared_ptr<dto::Schema>, std::shared_ptr<shd::Schema>>>(std::move(result.status), std::move(result.schema), std::move(shdSchemaPtr));
+        });
+}
+
+HTTPProxy::HTTPProxy() : _client(K23SIClientConfig()) {
+}
+
+seastar::future<> HTTPProxy::gracefulStop() {
+    _txns.clear();
```
Comment (ahsank) [inline src/k2/httpproxy/HTTPProxy.cpp:398]: Should it abort the txns like before?

Comment (iafuture) [reply src/k2/httpproxy/HTTPProxy.cpp:398]: This functionality should be implemented by the k23si client itself - when it shuts down, it should cleanup any in-progress txns. There is no reason why the user of the client library has to worry about it.

**Review by ahsank — APPROVED** [review]

File: src/k2/httpproxy/HTTPProxy.cpp [hunk 2]
```diff
@@ -476,92 +245,223 @@ HTTPProxy::_handleCreateSchema(sh::dto::CreateSchemaRequest&& request) {
...
+            auto& [status, k2Schema, shdSchema] = schemas;
+            if (!status.is2xxOK()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Status{.code=status.code, .message=status.message}, shd::WriteResponse{});
+            }
+            auto it = _txns.find(request.timestamp);
+            if (it == _txns.end()) {
+                return MakeHTTPResponse<shd::WriteResponse>(sh::Statuses::S410_Gone("transaction does not exist"), shd::WriteResponse{});
+            }
+            dto::SKVRecord k2record(request.collectionName, k2Schema);
+            bool haveKeys = !request.isDelete && request.fieldsForPartialUpdate.size() == 0;
```
Comment (jfunston) [inline src/k2/httpproxy/HTTPProxy.cpp:291]: We need the keys even if it is a delete or partial update in order to route. The K2 client will also verify the keys are present in the k2record as part of the write, so we can probably just pass true to the constructor.

File: src/k2/httpproxy/HTTPProxy.cpp [hunk 3]
```diff
@@ -476,92 +245,223 @@ HTTPProxy::_handleCreateSchema(sh::dto::CreateSchemaRequest&& request) {
...
+                }
+                auto it = _txns.find(request.timestamp);
+                if (it == _txns.end()) {
+                    return MakeHTTPResponse<shd::ReadResponse>(sh::Statuses::S410_Gone("transaction does not exist"), shd::ReadResponse{});
+                }
+                dto::SKVRecord k2record(request.collectionName, k2Schema);
+                shd::SKVRecord shdrecord(request.schemaName, shdSchema, std::move(request.key), true);
+                _shdRecToK2(shdrecord, k2record);
+
+                return it->second.handle.read(k2record.getKey(), request.collectionName)
```
Comment (jfunston) [inline src/k2/httpproxy/HTTPProxy.cpp:322]: Should we use the SKVRecord read instead?

**Review by jfunston — CHANGES_REQUESTED** [review]

File: src/k2/httpproxy/HTTPProxy.cpp (hunk 2 above)
Comment (iafuture) [reply src/k2/httpproxy/HTTPProxy.cpp:291]: good point. i'll just pass true

File: src/k2/httpproxy/HTTPProxy.cpp (hunk 3 above)
Comment (iafuture) [reply src/k2/httpproxy/HTTPProxy.cpp:322]: Yup, makes sense. no idea why i didn't do that in the first place :)

**Review by jfunston — APPROVED** [review]

File: src/k2/httpproxy/HTTPProxy.cpp [hunk 4]
```diff
@@ -395,9 +395,17 @@ HTTPProxy::HTTPProxy() : _client(K23SIClientConfig()) {
...
+    std::vector<seastar::future<>> futs;
+    for (auto& [ts, txn]: _txns) {
+        futs.push_back(txn.handle.kill());
+    }
+    return seastar::when_all_succeed(futs.begin(), futs.end()).discard_result()
+    .then([this] {
+        _txns.clear();
+    })
+    .then([this] {
+        return _client.gracefulStop();
```
Comment (ahsank) [inline src/k2/httpproxy/HTTPProxy.cpp:406]: What happens if seastar calls `K23SIClient::gracefulStop()` before calling HttpProxy `gracefulStop`? Then _client gracefulstop will be called twice.

**Review by ahsank — APPROVED** [review]

File: src/k2/httpproxy/HTTPProxy.cpp (hunk 4 above)
Comment (jfunston) [reply src/k2/httpproxy/HTTPProxy.cpp:406]: We are not using the K23SIClient as an applet, so gracefulStop is not called for it automatically. Instead the HTTPProxy owns the client directly (it is a member variable) so it is responsible for starting and stopping it.

Comment (ahsank) [reply src/k2/httpproxy/HTTPProxy.cpp:406]: Yes forgot that it's just client api not applet.

**Pull request merged**